addons/controllers/clusterbootstrap_controller.go

```diff
@@ -284,62 +284,195 @@ func (r *ClusterBootstrapReconciler) createOrPatchClusterBootstrapFromTemplate(c
+	clusterBootstrap.Name = cluster.Name
+	clusterBootstrap.Namespace = cluster.Namespace
+	clusterBootstrap.Spec = clusterBootstrapTemplate.Spec.DeepCopy()
 
-		clusterBootstrap.OwnerReferences = []metav1.OwnerReference{
-			{
-				APIVersion:         clusterapiv1beta1.GroupVersion.String(),
-				Kind:               cluster.Kind,
-				Name:               cluster.Name,
-				UID:                cluster.UID,
-				Controller:         pointer.BoolPtr(true),
-				BlockOwnerDeletion: pointer.BoolPtr(true),
-			},
-		}
+	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+		clusterBootstrap.Spec.CNI,
+		clusterBootstrap.Spec.CPI,
+		clusterBootstrap.Spec.CSI,
+		clusterBootstrap.Spec.Kapp,
+	}, clusterBootstrap.Spec.AdditionalPackages...)
 
-		if err := r.Client.Create(r.context, clusterBootstrap); err != nil {
-			return nil, err
-		}
-		// ensure ownerRef of clusterBootstrap on created secrets and providers, this can only be done after
-		// clusterBootstrap is created
-		ownerRef := metav1.OwnerReference{
-			APIVersion:         runtanzuv1alpha3.GroupVersion.String(),
-			Kind:               "ClusterBootstrap", // kind is empty after create
-			Name:               clusterBootstrap.Name,
-			UID:                clusterBootstrap.UID,
+	secrets, providers, err := r.cloneSecretsAndProvidersFromPackageList(cluster, packages, clusterBootstrapTemplate.Namespace, log)
+	if err != nil {
+		r.Log.Error(err, "unable to clone secrets or providers")
+		return nil, err
+	}
+
+	clusterBootstrap.OwnerReferences = []metav1.OwnerReference{
+		{
+			APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+			Kind:               cluster.Kind,
+			Name:               cluster.Name,
+			UID:                cluster.UID,
 			Controller:         pointer.BoolPtr(true),
 			BlockOwnerDeletion: pointer.BoolPtr(true),
+		},
+	}
+
+	if err := r.Client.Create(r.context, clusterBootstrap); err != nil {
+		return nil, err
+	}
+	// ensure ownerRef of clusterBootstrap on created secrets and providers, this can only be done after
+	// clusterBootstrap is created
+	if err := r.ensureOwnerRef(clusterBootstrap, secrets, providers); err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to ensure ClusterBootstrap %s/%s as a ownerRef on created secrets and providers", clusterBootstrap.Namespace, clusterBootstrap.Name))
+		return nil, err
+	}
+
+	clusterBootstrap.Status.ResolvedTKR = tkrName
+	if err := r.Status().Update(r.context, clusterBootstrap); err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to update the status of ClusterBootstrap %s/%s", clusterBootstrap.Namespace, clusterBootstrap.Name))
+		return nil, err
+	}
+	r.Log.Info("cloned clusterBootstrap", "clusterBootstrap", clusterBootstrap)
+
+	return clusterBootstrap, nil
+}
+
+// patchClusterBootstrapFromTemplate will patch ClusterBootstrap associated with a cluster in case of TKR upgrade
+func (r *ClusterBootstrapReconciler) patchClusterBootstrapFromTemplate( // nolint:funlen,gocyclo
```

Comment:
Just noticed these, please refactor to remove lint suppression.

Reply:
ack

Reply:
Hi @vijaykatam , my recent commit have addressed the lint issue